ulib/experimental/Steel.Reference.fsti

```diff
@@ -198,20 +222,72 @@ val free (#a:Type0) (r:ref a) : Steel unit
   (ensures fun _ _ _ -> True)
 
 /// Reads the current value of reference [r]
-val read (#a:Type0) (r:ref a) : Steel a
+val readp (#a:Type0) (r:ref a) (p: perm) : Steel a
+  (vptrp r p) (fun _ -> vptrp r p)
+  (requires fun _ -> True)
+  (ensures fun h0 x h1 -> h0 (vptrp r p) == h1 (vptrp r p) /\ x == h1 (vptrp r p))
+
+let read (#a:Type0) (r:ref a) : Steel a
   (vptr r) (fun _ -> vptr r)
   (requires fun _ -> True)
   (ensures fun h0 x h1 -> sel r h0 == sel r h1 /\ x == sel r h1)
+= readp r full_perm
 
 /// Writes value [x] in reference [r]
 val write (#a:Type0) (r:ref a) (x:a) : Steel unit
   (vptr r) (fun _ -> vptr r)
   (requires fun _ -> True)
   (ensures fun _ _ h1 -> x == sel r h1)
 
+val share (#a:Type0) (#uses:_) (#p: perm) (r:ref a)
+  : SteelGhost unit uses
+    (vptrp r p)
+    (fun _ -> vptrp r (half_perm p) `star` vptrp r (half_perm p))
+    (fun _ -> True)
+    (fun h _ h' ->
+      h' (vptrp r (half_perm p)) == h (vptrp r p)
+    )
+
+val gather_gen (#a:Type0) (#uses:_) (r:ref a) (p0:perm) (p1:perm)
+  : SteelGhost perm uses
+    (vptrp r p0 `star` vptrp r p1)
+    (fun res -> vptrp r res)
+    (fun _ -> True)
+    (fun h res h' ->
+      res == sum_perm p0 p1 /\
+      h' (vptrp r res) == h (vptrp r p0) /\
+      h' (vptrp r res) == h (vptrp r p1)
+    )
+
+let gather (#a: Type0) (#uses: _) (#p: perm) (r: ref a)
+  : SteelGhost unit uses
+      (vptrp r (half_perm p) `star` vptrp r (half_perm p))
```

Comment:
Is the specialization needed? Couldn't we just have the generic `gather_gen` (renamed into `gather`)? Is bidirectional framing correctly solving permission implicits with this version?

Reply:
Bidirectional framing correctly infers the arguments to `gather`, as you expected; however the problem lies more with its return permission. Indeed, with:
```
val gather (#a:Type0) (#uses:_)  (#p0:perm) (#p1:perm) (r:ref a)
  : SteelGhost unit uses
    (vptrp r p0 `star` vptrp r p1)
    (fun res -> vptrp r ( (p0 `sum_perm` p1)))
    (fun _ -> True)
    (fun h res h' ->
      h' (vptrp r ( (p0 `sum_perm` p1))) == h (vptrp r p0) /\
      h' (vptrp r ( (p0 `sum_perm` p1))) == h (vptrp r p1)
    )
```
`Selectors.Examples.test4` above fails with:
```
Goal 1/1 (implicit var for binder (#p2:
Prims.squash (Steel.Effect.Common.can_be_split_post #a
      #b
      (fun x y -> Steel.Effect.Common.star (post_g x y) (frame_g x))
      post)) of (SteelGhostBase, SteelGhostBase) |> SteelGhostBase at <input>(67,2-67,9))
r: ref nat
x x'0: unit
--------------------------------------------------------------------------------
squash (equiv (vptrp #nat r Steel.FractionalPermission.full_perm)
      (VStar (vptrp #nat r (Steel.FractionalPermission.sum_perm (*?u672*) _ (*?u671*) _))
          ((*?u670*) _ x)))
(*?u675*) _

Failed to resolve implicit argument ?16 of type Steel.FractionalPermission.perm introduced for Instantiating implicit argument in application
```
because, as I feared, `sum_perm (half_perm full_perm) (half_perm full_perm)` will not unify with `full_perm`, and I need to add an explicit `change_equal_slprop (vptrp r _) (vptr r)` at the end of `test4`.
Adding `normal` to the return permission of `gather` does not really help, with `test4` failing with a more meaningful message:
```
(Error 228) user tactic failed: `could not find a solution for unifying
Steel.Reference.vptrp #Prims.nat r Steel.FractionalPermission.full_perm
and
Steel.Reference.vptrp #Prims.nat
  r
  (Steel.Effect.Common.normal #Steel.FractionalPermission.perm
      (Steel.FractionalPermission.sum_perm (*?u672*) _ (*?u671*) _)) * (*?u670*) _ x`
```
From what I remember, `Steel.FractionalPermission` defines permissions as boxed real numbers because F* reflection currently does not support real number literals, so I don't think there is much to hope for there.
That is why, if you mean that we should make the common case work as smoothly as possible, and if the common case is to split permissions in half and gather them in the same order as the one in which they were split, then the signature of `gather` has to be based on unifying with two `half_perm p`.

Reply:
Ah, that makes sense. Feel free to merge then